dpbench/infrastructure/benchmark_runner.py

```diff
@@ -0,0 +1,403 @@
+# SPDX-FileCopyrightText: 2022 - 2023 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+"""Provides infrastructure to run benchmarks."""
+
+import inspect
+import logging
+import multiprocessing as mp
+import multiprocessing.connection as mpc
+import os
+from dataclasses import dataclass
+
+import numpy as np
+import sqlalchemy
+
+import dpbench.config as cfg
+from dpbench.infrastructure.benchmark import Benchmark, _exec, _setup_func
+from dpbench.infrastructure.benchmark_results import BenchmarkResults
+from dpbench.infrastructure.datamodel import store_results
+from dpbench.infrastructure.enums import ErrorCodes, ValidationStatusCodes
+from dpbench.infrastructure.frameworks import Framework
+from dpbench.infrastructure.frameworks.fabric import build_framework
+from dpbench.infrastructure.runner import _print_results
+
+"""
+Send on process creation:
+    Framework configuration: so the process can set it up.
+
+Send on benchmark run request:
+    Benchmark configuration
+
+Send on benchmark finish:
+    Benchmark execution results
+"""
+
+
+@dataclass
+class BaseRunConfig:
+    """Run configuration that is required to run benchmark."""
+
+    benchmark: cfg.Benchmark
+    implementation: str
+    ref_framework: cfg.Framework = None
+    preset: str = "S"
+    repeat: int = 10
+    validate: bool = True
+    precision: str = None
+    print_results: bool = True
+
+    @classmethod
+    def from_instance(cls, instance):
+        """Creates BaseRunConfiguration from it's inheritor."""
+        return cls(
+            **{
+                k: v
+                for k, v in instance.__dict__.items()
+                if k in inspect.signature(cls).parameters
+            }
+        )
+
+
+@dataclass
+class RunConfig(BaseRunConfig):
+    """Extended run configuration that controls benchmark execution."""
+
+    conn: sqlalchemy.Engine = None
+    framework: cfg.Framework = None
+    timeout: float = 200.0
+    run_id: int = None
+    skip_expected_failures: bool = False
+
+
+class BenchmarkRunner:
+    """Benchmark runner that delegates runs between processes.
+
+    It creates process for each framework to avoid conflicts in framework
+    imports.
+    """
+
+    def __init__(
+        self,
+        method: str = "spawn",
+    ) -> None:
+        """Creates BenchmarkRunner. No processes get spawn at this point.
+
+        Args:
+            method: method for sub process creations. Currently only spawn is
+                supported. We need to get rid of GLOBAL config to get support
+                for 'fork' method.
+        """
+        self._ctx = mp.get_context(method)
+        self._framework_processes: dict[
+            str, tuple[mp.Process, mpc.Connection]
+        ] = {}
+
+    def get_process(
```

Comment:
It will be a better design to convert the `get_process` into a contextmanager where in the `kill_process` and any other cleanup is done on exit. Invariably, someone down the road will forget a corresponding `kill_process`. But can be follow up work.